sdk/rust/src/lib.rs

```diff
@@ -0,0 +1,499 @@
+
+        if http_url.scheme() != "http" && http_url.scheme() != "https" {
+            return Err(ClientError::InvalidHttpUrl("Invalid scheme".to_string()));
+        }
+
+        if ws_url.scheme() != "ws" && ws_url.scheme() != "wss" {
+            return Err(ClientError::InvalidWsUrl("Invalid scheme".to_string()));
+        }
+
+        Ok(Self {
+            http_url,
+            ws_url,
+            api_key: config.api_key,
+            client: reqwest::Client::new(),
+        })
+    }
+
+    pub async fn connect_websocket(&self) -> Result<WsClient, ClientError> {
+        let url_string = format!(
+            "{}{}{}",
+            self.ws_url.as_str().trim_end_matches("/"),
+            api_types::Route::V1.as_ref(),
+            api_types::Route::Ws.as_ref()
+        );
+        let mut request = url_string
+            .into_client_request()
+            .map_err(|e| ClientError::WsConnectFailed(e.to_string()))?;
+        if let Some(api_key) = self.api_key.clone() {
+            let bearer_token = format!("Bearer {}", api_key);
+            request.headers_mut().insert(
+                "Authorization",
+                bearer_token.parse().map_err(|_| {
+                    ClientError::WsConnectFailed("Failed to parse api key".to_string())
+                })?,
+            );
+        }
+        let (ws_stream, _) = connect_async(request)
+            .await
+            .map_err(|e| ClientError::WsConnectFailed(e.to_string()))?;
+
+        let (request_sender, request_receiver) = mpsc::unbounded_channel();
+        let (update_sender, update_receiver) = broadcast::channel(1000);
+
+        Ok(WsClient {
+            inner: Arc::new(WsClientInner {
+                request_sender,
+                update_receiver,
+                request_id: RwLock::new(0),
+                ws: tokio::spawn(WsClient::run(ws_stream, request_receiver, update_sender)),
+            }),
+        })
+    }
+
+    pub async fn get_opportunities(
+        &self,
+        params: Option<GetOpportunitiesQueryParams>,
+    ) -> Result<Vec<Opportunity>, ClientError> {
+        self.send(Route::GetOpportunities, params).await
+    }
+
+    pub async fn new_bid<T: OpportunityTrait>(
+        opportunity: T,
+        params: T::Params,
+        private_key: String,
+    ) -> Result<api_types::bid::BidCreate, ClientError> {
+        T::new_bid(opportunity, params, private_key)
+    }
+}
+
+pub trait OpportunityTrait {
+    type Params;
+
+    fn new_bid(
+        opportunity: Self,
+        params: Self::Params,
+        private_key: String,
+    ) -> Result<api_types::bid::BidCreate, ClientError>;
+}
+
+impl OpportunityTrait for api_types::opportunity::OpportunityEvm {
+    type Params = BidParamsEvm;
+
+    fn new_bid(
+        opportunity: Self,
+        bid_params: Self::Params,
+        private_key: String,
+    ) -> Result<BidCreate, ClientError> {
+        let private_key = private_key.parse::<LocalWallet>().map_err(|e| {
```

Comment:
I think parsing should happen sooner and not in this function.

Reply:
Do you think it's a good practice to store private_keys for both evm and svm inside the client?

Reply:
nope, this should be on the simple searcher side

Reply:
I see that you added parsing on simple searcher side, then we can change this argument to LocalWallet. Maybe add another associated type for this argument inside `OpportunityTrait`